### PR TITLE
API diff between .NET Core 2.0 and .NET Core 2.1

### PR DESCRIPTION
**Pull request by terrajobst**
This is meant as a final pass to make sure we haven't missed anything.

- [x] Merge comments from #62 
- [x] Assign area owners

### CODE REVIEW

File: 2018/NetCore20-vs-NetCore21-Diff/README.md [hunk 1]
```diff
@@ -0,0 +1,1893 @@
+# API diff between .NET Core 2.0 and .NET Core 2.1
+
+To the best of my knowledge this is an accureate diff between .NET Core 2.0 and
```
Comment (jnm2) [inline 2018/NetCore20-vs-NetCore21-Diff/README.md]: Typo `accureate`

File: 2018/NetCore20-vs-NetCore21-Diff/README.md [hunk 2]
```diff
@@ -0,0 +1,1893 @@
...
+.NET Core 2.1, excluding ASP.NET Core as well as the Windows Compaptibility
```
Comment (jnm2) [inline 2018/NetCore20-vs-NetCore21-Diff/README.md]: Typos `accureate` and `Compaptibility`

**Comment by terrajobst** [comment]
> +        bool System.IEquatable<System.IntPtr>.Equals(IntPtr value);

Why did we change it?

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:179 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        public static implicit operator Memory<T> (ArraySegment<T> arraySegment);

should be `segment`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:205 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        public static implicit operator ReadOnlyMemory<T> (ArraySegment<T> arraySegment);

should be `segment`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:320 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        public static implicit operator ReadOnlySpan<T> (ArraySegment<T> arraySegment);

should be `segment`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:348 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        public static implicit operator Span<T> (ArraySegment<T> arraySegment);

should be `segment`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:408 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        protected internal abstract bool TryGetArray(out ArraySegment<T> arraySegment);

should be `segment`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:568 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        public static implicit operator Memory<T> (ArraySegment<T> arraySegment);

should be `segment`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:205 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        bool System.IEquatable<System.UIntPtr>.Equals(UIntPtr value);

`other`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:482 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
>  namespace System.Buffers {

Should we revisit the contents of this namespace? It seems we have a bunch of types that are plumbing that should be in a nested namespace, such as `IRetainable`.

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:519 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +    public delegate void ReadOnlySpanAction<T, in TArg>(ReadOnlySpan<T> span, TArg arg); {

Looks like plumbing. Should be in a sub namespace.

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:612 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +    public delegate void SpanAction<T, in TArg>(Span<T> span, TArg arg); {

Looks like plumbing. Should be in a sub namespace.

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:618 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        public static bool TryWriteInt16BigEndian(Span<byte> destination, short value);

All `TryWriteXxx` should use `destination`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:680 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        public static void ZeroMemory(Span<byte> buffer);

`destination`

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:1663 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
> +        public static bool FixedTimeEquals(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right);

Should the parameters be named `span`/`other` or `left`/`right`? These are statics hence I'm leaning towards the latter.

---
Refers to: 2018/NetCore20-vs-NetCore21-Diff/README.md:1662 in 9b5c027. [](commit_id = 9b5c0274be056047231fbf5dad321e85708196cc, deletion_comment = False)

**Comment by terrajobst** [comment]
@stephentoub 

> > Looks like plumbing. Should be in sub namespace.
>
> @terrajobst, can you elaborate?  These are essentially just `Action<..>`, but because span can't be used as a generic type parameter, we had to create custom delegates where we needed delegates that take span, e.g. `String.Create`.
> 
> If that's all understood and the agreement was still that they should be in a subnamespace, what namespace?

@KrzysztofCwalina brought it up that we [should do another pass](https://github.com/dotnet/apireviews/pull/63#issuecomment-375418439) on the `System.Buffers`. I understand why these delegates are needed but they are currently used by a single ctor on `string` and seem to be more advanced and thus rarely used.

**Comment by danmosemsft** [comment]
Not @ahsonkhan probably @maryamariyan

File: 2018/NetCore20-vs-NetCore21-Diff/README.md [hunk 3]
```diff
@@ -0,0 +1,1893 @@
...
++    public static class DbProviderFactories {
++        public static DbProviderFactory GetFactory(DataRow providerRow);
++        public static DbProviderFactory GetFactory(DbConnection connection);
++        public static DbProviderFactory GetFactory(string providerInvariantName);
++        public static DataTable GetFactoryClasses();
++        public static IEnumerable<string> GetProviderInvariantNames();
++        public static void RegisterFactory(string providerInvariantName, DbProviderFactory factory);
++        public static void RegisterFactory(string providerInvariantName, string factoryTypeAssemblyQualifiedName);
++        public static void RegisterFactory(string providerInvariantName, Type providerFactoryClass);
++        public static
```
Comment (ahsonkhan) [inline 2018/NetCore20-vs-NetCore21-Diff/README.md]: Should we consider changing this to chars and become inconsistent with the array-based Read method but consistent with the BinaryWriter Write methods?

Comment (ahsonkhan) [inline 2018/NetCore20-vs-NetCore21-Diff/README.md]: Why were these parameters renamed? It looks like they were done as far back as https://github.com/dotnet/corefx/commit/e3fc18c58396737ee5e6bc80acf508031d63cc8e#diff-b66ed9c9f50e26cf2e4afaa99ec250b8.

These already shipped as the following in .NET Core 2.0:
```C#
public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback callback, object state);
public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object state);
```

cc @JeremyKuhne

Comment (ahsonkhan) [inline 2018/NetCore20-vs-NetCore21-Diff/README.md]: Inconsistent parameter naming here.

cc @ViktorHofer, @ianhays

Comment (ahsonkhan) [inline 2018/NetCore20-vs-NetCore21-Diff/README.md]: buffer?

Comment (ahsonkhan) [inline 2018/NetCore20-vs-NetCore21-Diff/README.md]: inputBuffer, outputBuffer?

Comment (ahsonkhan) [inline 2018/NetCore20-vs-NetCore21-Diff/README.md]: other? span?

Existing API is `public bool Equals(StringBuilder sb);`

Comment (JeremyKuhne) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: This one was a bug. We renamed from the original on desktop. "array" is the right parameter name.

Comment (JeremyKuhne) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: I would keep consistent with this class and not change from buffer.

**Comment by bartonjs** [comment]
It's not obvious how to reply to any of the line item comments.

Crypto:

> public static void ZeroMemory(Span<byte> buffer);

We could probably do better than "buffer", but "destination" seems wrong (it's not the output of any serialization or computation.  In discussion with Krzysztof it seems like we could rename it to "span", but that's probably not enough value-add to do it at this point.

> public static bool FixedTimeEquals(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right);

Left and right are what Krzysztof came up with when prompted.  I agree they're better than "span/other" since this is a static non-extension.  The best comparison is IEquatable\<T\>, which uses "x/y", but we don't like that, either.  So "left/right" still seems like the best for static equals methods.

File: 2018/NetCore20-vs-NetCore21-Diff/README.md (hunk 3 above)
Comment (bartonjs) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: This seems the same to me as the Brotli(etc) compress or decompress, which are `(source, destination)`.  So I think we would rename this only if we were renaming it on compression.  Or am I missing something?

Comment (maryamariyan) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: OK. I am assuming it can be kept as `(source, destination)` in both places.

Comment (maryamariyan) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: I'll change buffer in `BeginRead(byte[] buffer, ...)` to array to make consistent with the rest.

Comment (maryamariyan) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: based on the response below by @bartonjs, I'll keep it as is.

Comment (terrajobst) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: AFAICT these APIs aren't net-new. These are ports from .NET Framework. Changing them isn't an option.

Comment (terrajobst) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: The diff is messed up. No idea what's happening on GitHub...

**Comment by terrajobst** [comment]
Apparently GitHub diffs aren't the most compelling way to track this as they comments aren't really pointing to the correct line anymore.

Besides the namespace, are there are any open issues? @maryamariyan, I assume you thumbed up the ones you took care of, right?

File: 2018/NetCore20-vs-NetCore21-Diff/README.md (hunk 3 above)
Comment (ahsonkhan) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: Yes, the line numbers don't line up. Maybe because `large diffs are not rendered by default`.

Comment (terrajobst) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: In my own code review tool they show up properly. Way to go GitHub...

Comment (terrajobst) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: Agreed, local consistency with members on the same existing type trumps consistency with members on other types.

Comment (terrajobst) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: 👍

Comment (terrajobst) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: I'd prefer `span`.

**Comment by maryamariyan** [comment]
@terrajobst yes I thumbed up the ones I took care of and the ones with hooray emoji are meant to show no action was required as it was already fixed or not needed change

**Comment by terrajobst** [comment]
> @KrzysztofCwalina brought it up that we [should do another pass](https://github.com/dotnet/apireviews/pull/63#issuecomment-375418439) on the `System.Buffers` namespace. I understand why these delegates are needed but they are currently only used by a single ctor on `string`; the question is whether they are likely becoming so popular that they should be in the feature's core namespace.

Synced with Krzysztof. We're out of time and we're happy enough that we'll leave it as-is.

File: 2018/NetCore20-vs-NetCore21-Diff/README.md (hunk 3 above)
Comment (maryamariyan) [reply 2018/NetCore20-vs-NetCore21-Diff/README.md]: For new BrotliStream BeginRead/BeginWrite/Write/Read/ReadAsync/ReadAsync APIs I made sure we use buffer to stay consistent with Stream in PR dotnet/corefx#28581.

**Pull request merged**